compiler/analyzer.ml

```diff
@@ -506,20 +503,28 @@ and check_fdecl env id f anon =
         check_params_type_mismatch func_env' (constrained_param' :: acc) 
           (List.tl func_param_types) tl in
         
-  let param_types = 
-    let typ = (VarMap.find id func_env.scope).s_type in
-    match typ with
-      | Func(func) -> func.param_types
-      | _ -> fdecl_reassign_error id typ in
-  let func_env, param_types' = check_params_type_mismatch func_env [] param_types param_ssids in
+  let param_types, return_typ =
+    let f_typ = (VarMap.find id func_env.scope).s_type in
+    match f_typ with
+      | Func(func) -> func.param_types, func.return_type
+      | _ -> fdecl_reassign_error id f_typ in
+  let func_env, param_types' =
+    check_params_type_mismatch func_env [] param_types param_ssids in
   
   (* Re-evaluate function return type to see if it has been constrained above *)
   let func_env, return = check_expr func_env f.return in
 
   (* Unconstrained function return types are not allowed *)
   let Sast.Expr(_, ret_type) = return in
-  if ret_type = Any || ret_type = List(Unconst) then 
-    fdecl_unconst_error id 
+  let ret_type = match ret_type with
+    | Any -> fdecl_unconst_error id
+    | Unconst -> fdecl_unconst_error id
+    | List(Any) -> fdecl_unconst_error id
+    | _ as typ -> typ in
```

Comment:
That actually gets caught by the `Unconst` case here, since the nested function gets recursively called into `check_fdecl` and will throw before we even try to return it.